### PR TITLE
[BEAM-11985] Python Bigtable - Implement IO Request Count metrics

### PR DESCRIPTION
**Pull request by benWize**
**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by benWize** [comment]
Hi, @ajamato would you help me to review this? 
I create a new class to override `flush` method in `MutationsBatcher` so I can get the rows response of `table.mutate_rows`.
Each row has a google.rpc.status_pb2.Status for some of these codes https://cloud.google.com/bigtable/docs/status-codes is still missing how I can handle the grpc status codes because service_call_metrics is expecting int httpStatus or string status, but here the `grpc` status is an integer (0-14).
https://github.com/apache/beam/blob/82f08b4dabc923151683f69cde71372755c7227c/sdks/python/apache_beam/io/gcp/bigtableio.py#L124

**Comment by ajamato** [comment]
> Hi, @ajamato would you help me to review this?
> I create a new class to override `flush` method in `MutationsBatcher` so I can get the rows response of `table.mutate_rows`.
> Each row has a google.rpc.status_pb2.Status for some of these codes https://cloud.google.com/bigtable/docs/status-codes is still missing how I can handle the grpc status codes because service_call_metrics is expecting int httpStatus or string status, but here the `grpc` status is an integer (0-14).
> https://github.com/apache/beam/blob/82f08b4dabc923151683f69cde71372755c7227c/sdks/python/apache_beam/io/gcp/bigtableio.py#L124

Is there is a method on the status object to get the string status "ok", "cancelled" "invalid_argument", etc. use that. (ServiceCallMetric will lowercase it)

If there is only a way to obtain the numbers:
https://cloud.google.com/bigtable/docs/status-codes
0 (OK)
1 (CANCELLED)
2 (UNKNOWN)
...

Then please write a helper function to lookup the number from a map and convert it to a string.
Lookup the string with that helper function before the call to service_metric.call()

i.e.
grpc_status_string = convert_to_grpc_status_string(status.code())
service_metric.call(grpc_status_string)

File: sdks/python/apache_beam/io/gcp/bigtableio.py [hunk 1]
```diff
@@ -87,14 +113,45 @@ def __setstate__(self, options):
...
+
+  def start_service_call_metrics(self, project_id, instance_id, table_id):
+    resource = resource_identifiers.BigtableTable(
+        project_id, instance_id, table_id)
+    labels = {
+        monitoring_infos.SERVICE_LABEL: 'BigTable',
+        # TODO(JIRA-11985): Add Ptransform label.
+        monitoring_infos.METHOD_LABEL: 'google.bigtable.v2.MutateRows',
+        monitoring_infos.RESOURCE_LABEL: resource,
+        monitoring_infos.BIGTABLE_PROJECT_ID_LABEL: self.
```
Comment (ajamato) [inline sdks/python/apache_beam/io/gcp/bigtableio.py:135]: Please use this style:
 monitoring_infos.BIGTABLE_PROJECT_ID_LABEL: (
    self.beam_options['project_id'],)

Comment (benWize) [reply sdks/python/apache_beam/io/gcp/bigtableio.py]: Thanks, I made both changes

**Comment by benWize** [comment]
Integration test to validate correct error handling
Screen Shot 2021-09-23 at 12 58 41

**Comment by benWize** [comment]
Run Python_PVR_Flink PreCommit

File: sdks/python/apache_beam/internal/metrics/metric.py [hunk 2]
```diff
@@ -223,3 +223,31 @@ def convert_to_canonical_status_string(self, status):
         http_status_code in http_to_canonical_gcp_status):
       return http_to_canonical_gcp_status[http_status_code]
     return str(http_status_code)
+
+  @staticmethod
+  def convert_to_grpc_status_string(grpc_status):
```
Comment (ajamato) [inline sdks/python/apache_beam/internal/metrics/metric.py:228]: please rename variable grpc_status -> grpc_status_code

Comment (ajamato) [reply sdks/python/apache_beam/internal/metrics/metric.py:228]: please rename function to:
bigtable_error_code_to_grpc_status_string

File: sdks/python/apache_beam/internal/metrics/metric.py [hunk 3]
```diff
@@ -223,3 +223,31 @@ def convert_to_canonical_status_string(self, status):
...
+    # type: (int) -> str
+
+    """Converts a grpc status to a canonical GCP status code string."""
```
Comment (ajamato) [inline sdks/python/apache_beam/internal/metrics/metric.py:231]: Replace this comment
"""
Converts the bigtable error code to a canonical GCP status code string.

This Bigtable client library is not using the canonical http status code values (i.e. https://cloud.google.com/apis/design/errors)"
Instead they are numbered using an enum with these values corresponding to each status code:
https://cloud.google.com/bigtable/docs/status-codes
"""

Please also add Args and Returns pydoc comment

**Review by ajamato — APPROVED** [review]

File: sdks/python/apache_beam/internal/metrics/metric.py (hunk 2 above)
Comment (benWize) [reply sdks/python/apache_beam/internal/metrics/metric.py]: Done

File: sdks/python/apache_beam/internal/metrics/metric.py (hunk 3 above)
Comment (benWize) [reply sdks/python/apache_beam/internal/metrics/metric.py]: Fixed

**Comment by amezcuam** [comment]
Hello, @kileys @ajamato the changes are done, could you please help us with the review? thanks

**Review by ajamato — APPROVED** [review]
looks good to me. Thanks for the PR

**Comment by ajamato** [comment]
@pabloem would you please merge?

**Comment by benWize** [comment]
Thank you all!

**Comment by rashidjeffreysky** [comment]
Hi everyone,

We are seeing errors which _possibly_ originate from this code:

`def write_mutate_metrics(self, rows):
    for status in rows:
      grpc_status_string = (
          ServiceCallMetric.bigtable_error_code_to_grpc_status_string(
              status.code))
      self.service_call_metric.call(grpc_status_string)
`

Error message:
`
File "/usr/local/lib/python3.6/site-packages/apache_beam/io/gcp/bigtableio.py", line 123, in write_mutate_metrics status.code))
AttributeError: 'NoneType' object has no attribute 'code' [while running 'Write event Bigtable row/ParDo(_BigTableWriteFn)-ptransform-58134']
`

Would you have any idea about this issue? 


In addition, the choice of parameter name (rows) is confusing. Perhaps something like responses or response_list ?
https://cloud.google.com/bigtable/docs/writing-data#batch
https://googleapis.dev/python/bigtable/1.3.0/_modules/google/cloud/bigtable/table.html#Table.mutate_rows

Thank you for your time.

Regards
Rashid

**Comment by rashidjeffreysky** [comment]
Hi @benWize , @ajamato 

We downgraded to Beam 2.33 which does not contain the new child class _MutationsBatcher(MutationsBatcher)

We restarted our jobs and left them running overnight. We are no longer seeing the issue which I mentioned in my previous comment.

Our team is somewhat new to Beam so I wish I could provide more insight. Perhaps another code review would be beneficial.

2.33 https://beam.apache.org/releases/pydoc/2.33.0/_modules/apache_beam/io/gcp/bigtableio.html#WriteToBigTable.expand
2.34 https://beam.apache.org/releases/pydoc/2.34.0/_modules/apache_beam/io/gcp/bigtableio.html#WriteToBigTable.expand
2.35 https://beam.apache.org/releases/pydoc/2.35.0/_modules/apache_beam/io/gcp/bigtableio.html#WriteToBigTable.expand

Many thanks.

Regards
Rashid

**Comment by ajamato** [comment]
Please be aware that I, @ajamato no longer work with Google or on Beam. You
can reach out to Ahmet if you need someone else to help complete the PR
review.

**Comment by aaltay** [comment]
@KevinGG - could you look at this reported issue? I think we can warn on issues related to reporting IO metrics instead of hard failing a pipeline.

**Comment by KevinGG** [comment]
> @KevinGG - could you look at this reported issue? I think we can warn on issues related to reporting IO metrics instead of hard failing a pipeline.

Filed BEAM-13602.

**Comment by KevinGG** [comment]
@rashidjeffreysky, @aaltay 
The expected behavior is a single response (based on the [REST API](https://cloud.google.com/bigtable/docs/reference/data/rpc/google.bigtable.v2#mutaterowsresponse)) with multiple `Status` entries. In the Python client implementation of bigtable, the returned response is a `List[Optional[Status]]` (or equivalently `List[Union[None, Status]]`)).  Details: default [value](https://github.com/googleapis/python-bigtable/blob/fec06fcd28c36d0d3b347b43d1f3d264e5f5aa39/google/cloud/bigtable/table.py#L1058) is None, retryable errors will [return default value](https://github.com/googleapis/python-bigtable/blob/fec06fcd28c36d0d3b347b43d1f3d264e5f5aa39/google/cloud/bigtable/table.py#L1078) instead of a `Status` instance)

I'll send a PR to do the following 2 things:

1. Prevent the metrics gathering from failing the job;
2. Handle None cases and treat them as ['DEADLINE_EXCEEDED'](https://grpc.github.io/grpc/core/md_doc_statuscodes.html).

I fear that there is data-loss throughout the process. If a row mutation fails but no error is raised, DF nor bigtable no longer retries.

**Comment by KevinGG** [comment]
Sent https://github.com/apache/beam/pull/16449 to fix BEAM-13602 and filed further TODO item BEAM-13606 for anyone who wants to improve the I/O to handle non-ok scenarios.

**Comment by aaltay** [comment]
Thank you!

**Pull request merged**